csharp/ql/src/semmle/code/csharp/controlflow/internal/Splitting.qll

```diff
@@ -1023,6 +1025,165 @@ module BooleanSplitting {
+  private import PreBasicBlocks
+  private import PreSsa
+
+  /** Holds if `ce` is guarded by a (non-)empty check, as specified by `v`. */
+  private predicate emptinessGuarded(
+    Guards::Guard g, Guards::CollectionExpr ce, Guards::AbstractValues::EmptyCollectionValue v
+  ) {
+    exists(PreBasicBlock bb | Guards::Internal::preControls(g, bb, v) |
+      PreSsa::adjacentReadPairSameVar(g, ce) and
+      bb.getAnElement() = ce
+    )
+  }
+
+  /**
+   * A loop where the body is guaranteed to be executed at least once, and
+   * can therefore be unrolled in the control flow graph.
+   */
+  abstract class UnrollableLoopStmt extends LoopStmt {
+    /** Holds if the step `pred --c--> succ` should start loop unrolling. */
+    abstract predicate startUnroll(ControlFlowElement pred, ControlFlowElement succ, Completion c);
+
+    /** Holds if the step `pred --c--> succ` should stop loop unrolling. */
+    abstract predicate stopUnroll(ControlFlowElement pred, ControlFlowElement succ, Completion c);
+
+    /**
+     * Holds if any step `pred --c--> _` should be pruned from the unrolled loop
+     * (the loop condition evaluating to `false`).
+     */
+    abstract predicate pruneLoopCondition(ControlFlowElement pred, ConditionalCompletion c);
+
+    /** Gets a descendant that belongs to the body of this loop. */
+    ControlFlowElement getABodyDescendant() {
+      result = this.getBody()
+      or
+      exists(ControlFlowElement mid |
+        mid = this.getABodyDescendant() and
+        result = getAChild(mid, mid.getEnclosingCallable())
+      )
+    }
+  }
+
+  private class UnrollableForeachStmt extends UnrollableLoopStmt, ForeachStmt {
+    UnrollableForeachStmt() {
+      exists(Guards::AbstractValues::EmptyCollectionValue v | v.isNonEmpty() |
+        emptinessGuarded(_, this.getIterableExpr(), v)
+        or
+        this.getIterableExpr() = v.getAnExpr()
+      )
+    }
+
+    override predicate startUnroll(ControlFlowElement pred, ControlFlowElement succ, Completion c) {
+      pred = last(this.getIterableExpr(), c) and
+      succ = this
+    }
+
+    override predicate stopUnroll(ControlFlowElement pred, ControlFlowElement succ, Completion c) {
+      pred = last(this.getBody(), c) and
+      succ = succ(pred, c) and
+      not succ = this.getABodyDescendant()
+    }
+
+    override predicate pruneLoopCondition(ControlFlowElement pred, ConditionalCompletion c) {
+      pred = this and
+      c.(EmptinessCompletion).isEmpty()
+    }
+  }
+
+  /**
+   * A split for loops where the body is guaranteed to be executed at least once, and
+   * can therefore be unrolled in the control flow graph. For example, in
+   *
+   * ```
+   * void M(string[] args)
+   * {
+   *     if (args.Length == 0)
+   *         return;
+   *     foreach (var arg in args)
+   *         System.Console.WriteLine(args);
+   * }
+   * ```
+   *
+   * the `foreach` loop is guaranteed to be executed at least once, as a result of the
+   * `args.Length == 0` check.
```

Comment:
`args.Length == 0` seems a little specific. Perhaps "an `EmptinessCompletion` check"?

Reply:
It is just an example, so isn't that ok?

Reply:
That's fine I just misread the comment.